### PR TITLE
PCRE matching of string fields

### PR DESCRIPTION
**Pull request by Totktonada**
Related to #73.

### CODE REVIEW

File: graphql/accessor_general.lua [hunk 1]
```diff
@@ -696,6 +702,28 @@ local function validate_collections(collections, schemas)
...
+--- XXX
+local function match_using_re(obj, pcre)
+    if pcre == nil then return true end
+
+    for field_name, re in pairs(pcre) do
+        -- skip an object with null in a string* field
+        if obj[field_name] == nil then
+            return false
+        end
+        assert(rex ~= nil, 'we should not pass over :compile() ' ..
```
Comment (Khatskevich) [inline graphql/accessor_general.lua:722]: Is this comment or assert msg?)

File: graphql/accessor_general.lua [hunk 2]
```diff
@@ -696,6 +702,28 @@ local function validate_collections(collections, schemas)
     end
 end
 
+--- XXX
+local function match_using_re(obj, pcre)
+    if pcre == nil then return true end
```
Comment (Khatskevich) [inline graphql/accessor_general.lua]: Is this situation normal?
I suppose assert is more suitable here and this function is called only if necessary. But it is up to you...

File: graphql/accessor_general.lua [hunk 3]
```diff
@@ -741,9 +769,11 @@ local function process_tuple(state, tuple, opts)
                     qstats.fetched_object_cnt, fetched_object_cnt_max))
     assert(qcontext.deadline_clock > clock.monotonic64(),
            'query execution timeout exceeded, use `timeout_ms` to increase it')
+    local collection_name = opts.collection_name
```
Comment (Khatskevich) [inline graphql/accessor_general.lua:780]: I do not like creating variables for one use.

File: graphql/accessor_general.lua [hunk 4]
```diff
@@ -696,6 +702,28 @@ local function validate_collections(collections, schemas)
     end
 end
 
+--- XXX
+local function match_using_re(obj, pcre)
```
Comment (Khatskevich) [inline graphql/accessor_general.lua]: I prefer `regexp` name over `pcre`. 
Pcre dis
1. can be confusing for newcomers
2. emphasizes implementation.

File: graphql/accessor_general.lua [hunk 5]
```diff
@@ -828,6 +859,8 @@ local function select_internal(self, collection_name, from, filter, args, extra)
     -- XXX: save type at parsing and check here
     --assert(args.offset == nil or type(args.offset) == 'number',
     --    'args.offset must be a number of nil, got ' .. type(args.offset))
+    assert(args.pcre == nil or type(args.pcre) == 'table',
```
Comment (Khatskevich) [inline graphql/accessor_general.lua:870]: I do not like, that we loose structure of filters.
I suggest:
1. Create single attribute `filter`
2. `filter` is a table which contains
    - `regexp`
    - `brute` (i do not insist on the naming, I insist on sense)
    - `index`

I believe that this approach
1. much more extensible 
2. would speed up future coding
3. do not ban such names of fields like `pcre`

File: graphql/accessor_general.lua [hunk 6]
```diff
@@ -976,6 +1010,87 @@ local function init_qcontext(accessor, qcontext)
...
+local function get_primary_key_type(self, collection_name)
+    -- get name of field of primary key
+    local _, index_meta = get_primary_index_meta(
+        self, collection_name)
+
+    local offset_fields = {}
+
+    for _, field_name in ipairs(index_meta.fields) do
+        local field_type
+        local collection = self.collections[collection_name]
```
Comment (Khatskevich) [inline graphql/accessor_general.lua]: Please, initialize variables which are not related to the particular loop cycle outside of the loop. It is a little coonfusing.

File: graphql/accessor_general.lua [hunk 7]
```diff
@@ -976,6 +1010,87 @@ local function init_qcontext(accessor, qcontext)
...
+        local schema = self.schemas[collection.schema_name]
+        for _, field in ipairs(schema.fields) do
+            if field.name == field_name then
```
Comment (Khatskevich) [inline graphql/accessor_general.lua]: What is about rested records?

File: graphql/accessor_general.lua [hunk 8]
```diff
@@ -976,6 +1010,87 @@ local function init_qcontext(accessor, qcontext)
         settings.timeout_ms * 1000 * 1000
 end
 
+--- XXX
+local function get_primary_key_type(self, collection_name)
```
Comment (Khatskevich) [inline graphql/accessor_general.lua]: rename pk_avro_type

File: graphql/accessor_general.lua [hunk 9]
```diff
@@ -976,6 +1010,87 @@ local function init_qcontext(accessor, qcontext)
...
+end
+
+-- XXX: add string fields of a nested record / 1:1 connection
+
+--- XXX
+--- Note: it is only for list of objects and is not for destination of 1:1
+--- connection
+local function get_pcre_argument_type(self, collection_name)
+    local collection = self.collections[collection_name]
+    assert(collection ~= nil, 'cannot found collection ' ..
```
Comment (Khatskevich) [inline graphql/accessor_general.lua]: Please do not write long messages to errors which should not occur. It just makes code-base huge and increases maintenance costs.

File: graphql/accessor_general.lua [hunk 10]
```diff
@@ -1114,53 +1229,20 @@ function accessor_general.new(opts, funcs)
...
-                    offset_type = {
-                        name = collection_name .. '_offset',
-                        type = 'record',
-                        fields = offset_fields,
-                    }
+                -- add `pcre` argument only if lrexlib-pcre was found
+                local pcre_field
+                if rex ~= nil then
+                    local pcre_type = get_pcre_argument_type(self, collection_name)
+                    pcre_field = {name = 'pcre', type = pcre_type}
```
Comment (Khatskevich) [inline graphql/accessor_general.lua:1266]: I have wrote above. I do not lile `pcre` name. Moreover in user-visible part.

File: test/testdata/compound_index_testdata.lua [hunk 11]
```diff
@@ -459,6 +459,34 @@ function compound_index_testdata.run_queries(gql_wrapper)
...
+                    user_num: $user_num}) {
+                user_str
+                user_num
+                last_name
+                first_name
+            }
+        }
+    ]]
+
+    utils.show_trace(function()
```
Comment (Khatskevich) [inline test/testdata/compound_index_testdata.lua:478]: What is the purpose of this diff?

File: graphql/accessor_general.lua (hunk 1 above)
Comment (Totktonada) [reply graphql/accessor_general.lua:722]: Propose one better or I’ll leave it as is.

File: graphql/accessor_general.lua (hunk 2 above)
Comment (Totktonada) [reply graphql/accessor_general.lua]: `pcre` is like `filter`. You maybe confuse it with `rex` — module variable.

File: graphql/accessor_general.lua (hunk 3 above)
Comment (Totktonada) [reply graphql/accessor_general.lua:780]: I do not like unpacking some opts, but don’t unpacking others. Debatable and non-important. Going to leave as is.

File: graphql/accessor_general.lua (hunk 4 above)
Comment (Totktonada) [reply graphql/accessor_general.lua]: 1. I don’t know a developer that don’t know PCRE term.
2. It holds a standard, not an implementation. Consider comment with link to the alternative lua PCRE implementation.

I’m tentative about your proposal. Maybe.

File: graphql/accessor_general.lua (hunk 5 above)
Comment (Totktonada) [reply graphql/accessor_general.lua:870]: Not in the scope of this issue.

File: graphql/accessor_general.lua (hunk 6 above)
Comment (Totktonada) [reply graphql/accessor_general.lua]: Ok.

File: graphql/accessor_general.lua (hunk 7 above)
Comment (Totktonada) [reply graphql/accessor_general.lua]: XXX mark above. Will file an issue re nested records and 1:1 connections (nested filters).

File: graphql/accessor_general.lua (hunk 8 above)
Comment (Totktonada) [reply graphql/accessor_general.lua]: 1. Here all types are avro.
2. I prefer to omit abbreviations where possible.

Debatable, non-important. Going to leave as is.

File: graphql/accessor_general.lua (hunk 9 above)
Comment (Totktonada) [reply graphql/accessor_general.lua]: > cannot found collection %s

This error message is not long. You are about some other assert I guess. Propose your wording.

File: graphql/accessor_general.lua (hunk 10 above)
Comment (Totktonada) [reply graphql/accessor_general.lua:1266]: I’m going to think about this bunch of problems in the scope of #39. Not now.

File: test/testdata/compound_index_testdata.lua (hunk 11 above)
Comment (Totktonada) [reply test/testdata/compound_index_testdata.lua:478]: Do you read 215c1dfaa3ba7e24a83a2a3bbd3b052ba63bcc36 comment?

**Pull request merged**